### ISSUE TITLE
Jest and Mongoose not working well together

### ISSUE THREAD

**Issue by jozsi** [Automattic/mongoose#5033]
**EDIT** Repository to reproduce: https://github.com/jozsi/bug-jest-mongoose

Using the latest version of Jest (`19.0.2`) and Mongoose (`4.8.5`), the following code is breaking when ran by jest:
```js
const mongoose = require('mongoose');
const userSchema = new mongoose.Schema({
  email: {
    type: String,
    unique: true,
  },
});

const User = mongoose.model('User', userSchema);

User.on('index', function(err) {
  console.log('Index creation errored?', err);
});

mongoose.connect('localhost/test123');
```

Expected result (when ran by Node.js (`7.6.0`)):
```
$ node test.js
Index creation errored? undefined
```

Actual result (when ran by Jest):
```
$ ./node_modules/.bin/jest --no-cache test.js
 FAIL  ./test.js
  ● Test suite failed to run

    Your test suite must contain at least one test.
      
      at onResult (node_modules/jest/node_modules/jest-cli/build/TestRunner.js:192:18)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        1.277s
Ran all test suites matching "test.js".
  console.log test.js:13
    Index creation errored? Error {
      name: 'MongoError',
      message: 'Index keys cannot be empty.',
      ok: 0,
      errmsg: 'Index keys cannot be empty.',
      code: 67,
      codeName: 'CannotCreateIndex' }
```

I am using the default jest configuration (=> automocking is disabled).

Related: https://github.com/facebook/jest/issues/3045

**Comment by varunjayaraman** [comment]
I've noticed that Mongoose and Jest don't play together. I'm not sure if it's a mongoose thing or a Jest thing, but let me know what you find with the issue you opened up. There was another issue opened up recently relating to promises and I was wondering if Jest somehow monkey-patches `Promise`, but i can't imagine it would (and I would really hope it doesn't)

**Comment by jozsi** [comment]
@varunjayaraman - I am afraid it's not (just) jest and/or jsdom.

For instance, mongoose has the following [workaround](https://github.com/Automattic/mongoose/blob/master/lib/utils.js#L495-L521) that causes thrown errors to bypass any catch wrapper of jest => it just exists the process. I haven't found a workaround for that yet (tried the `__noPromises` flag to reduce wrappers, tried disabling `emitIndexErrors` to make sure the error is thrown only once, ...). I wonder if that portion is code is still actual with newer `node-mongodb-native` versions.
